### DISCUSSION TITLE
RAM and script size

### DISCUSSION THREAD

**Discussion by Novakasa** [pybricks/support#483] [Q&A] [answered]
I am currently running into memory limitations with my script running on a CityHub. As I am inexperienced working on such a low memory system, I have some questions:
- How much RAM does the CityHub and TechnicHub have for micropython programs?
- Does the available RAM get smaller if I have a larger script? Is the size of the script negligible compared to total RAM?
- What are some tools I can use to investigate memory issues?

Thanks for anyone who can shed some light onto this.

**Comment by laurensvalk** [comment, marked as answer]
You can see memory statistics with [mem_info](https://docs.pybricks.com/en/latest/micropython/micropython.html).

![image](https://user-images.githubusercontent.com/12326241/133587703-72d32f2a-a522-4662-927f-ed7b4e0f9928.png)

The available heap is configured [here](https://github.com/pybricks/pybricks-micropython/blob/e760080a9c9d4b6070ca771ad3a9c8d5021c9d7c/bricks/cityhub/mpconfigport.h#L11) for City Hub. You can find a similar file for Technic Hub.

> Does the available RAM get smaller if I have a larger script?

Yes, because the script is also held in RAM. You may be able to verify using the tools above.

> Is the size of the script negligible compared to total RAM?

Certainly not on Move Hub and City Hub. The other hubs have a bit more to spare.

**Reply by Novakasa** [reply]
Thanks, that helps a lot!

**Reply by Novakasa** [reply]
A followup question to this: If I were to include a large program with the firmware, would it further reduce available RAM for a dynamically started program?

**Reply by dlech** [reply]
No, the RAM usage will be the same for a program started from Pybricks Code vs. the same program started from flash memory.

**Reply by Novakasa** [reply]
Hi, sorry for the late reply, but could you rephrase this answer?
I guess my question is: Is the `main.py` also held in RAM?

If I split my program across a `main.py` and a dynamically started program, does this give me more available RAM, since the dynamically started program is smaller? (inspired by https://github.com/pybricks/support/issues/284#issuecomment-889840978)

**Reply by Novakasa** [reply]
Okay, I just stumbled upon the discussion in https://github.com/pybricks/support/issues/408#issuecomment-890322283, so it seems currently, the `main.py` will be copied into ram when importing it, but it might be possible in the future to avoid that.

**Reply by dlech** [reply]
I think the point of misunderstanding is that no matter where the `main.mpy` file is stored, it still has to be run through the MicroPython interpreter. This doesn't copy the file verbatim to RAM, but rather allocates function object for each function defined, type object for each class defined, etc. All of these objects take up about the same amount of RAM as the file. But there currently isn't a way to create static types or function in .mpy files like it is possible when writing MicroPython code in C. I found these open issues about this subject in MicroPython: https://github.com/micropython/micropython/issues/4067, https://github.com/micropython/micropython/issues/4073, https://github.com/micropython/micropython/issues/4124. They should do a better job of explaining it that I can.

**Reply by Novakasa** [reply]
Thank you, that clears it up. It would be amazing if we could save memory with this, but I get that this is an upstream issue.

**Reply by Novakasa** [reply]
But now I'm wondering, for a dynamically started program, we also have to allocate the memory for functions and classes. So in that case, we have to use memory for both these objects, as well as the mpy file itself. I guess the mpy file is deleted once parsed, so this doesn't use more memory than running the program as `main.py`?

**Reply by dlech** [reply]
> I guess the mpy file is deleted once parsed, so this doesn't use more memory than running the program as `main.py`?

Correct.
